### ISSUE TITLE
improve performance result

### ISSUE THREAD

**Issue by zhiqiangwan** [JamesChuanggg/vqa-tf#2]
Hi, I found that you missed the fully connected layer after the fusion of image feature and question feature. Besides, the original code in torch is implemented in multiple choice task. You used his pre-processing code and it means the data you used to train your model is  multiple choice task. But you said your accuracy is achieved in open-ended task.
I wonder whether these are the reasons causing the low accuracy.

**Comment by chingyaoc** [comment]
Hi, thanks for giving me lots of advice. The original code simply add a liner embedding layer after fusion, could you show the exact part of fully connected layer? Btw I think that pre-processing code give us image feature which is independent to the task type. I write this code based on [this](https://github.com/VT-vision-lab/VQA_LSTM_CNN/blob/master/train.lua) which is the original code released by VT.

**Comment by zhiqiangwan** [comment]
In his paper, there is a fully connected layer which is shown in the picture below.
![vqa](https://cloud.githubusercontent.com/assets/19935904/16358326/e6812310-3add-11e6-914f-c61c19d6ab5a.png)
The image feature is independent to the task type, but the questions are not. In this code https://github.com/JamesChuanggg/vqa-tf/blob/master/data/vqa_preprocessing.py#L50, the questions used to train your model is multiple choice task not open-ended task.

I also has some questions about your code in model_VQA.py.
1. https://github.com/JamesChuanggg/vqa-tf/blob/master/model_VQA.py#L322
why you add 1 to the vocabulary size?
2. https://github.com/JamesChuanggg/vqa-tf/blob/master/model_VQA.py#L43
why the number of LSTM states is 2 \* rnn_size \* rnn_layer? I think it should be rnn_size \* rnn_layer. I also check the BasicLSTMCell class in  rnn_cell and it multiply the states by 2. This is what I don't quite understand.

Thanks!

**Comment by chingyaoc** [comment]
Hi, my fc_layer is [here](https://github.com/JamesChuanggg/vqa-tf/blob/master/model_VQA.py#L114). Vqa_processing.py is a little bit weird, but I trained the original torch code with this data and the result was reasonable. Maybe we should open a issue [here](https://github.com/VT-vision-lab/VQA_LSTM_CNN)

About the questions, 
1. I removed "plus 1" in my latest version, sorry for confusing.
2. Since I concatenate output and hidden state just like the original code. The output size and hidden state size of LSTM are both rnn_size,  which will result in 2 \* rnn_size \* rnn_layer.

**Comment by zhiqiangwan** [comment]
Thank you for your reply. It makes a lot sense. 
https://github.com/JamesChuanggg/vqa-tf/blob/master/model_VQA.py#L114, I think this is just the linear transformation for softmax layer not the fc_layer. I have checked the author's original code. He did not add a fc_layer in his code.

**Comment by chingyaoc** [comment]
Hi, I've update the code and it can reach the same accuracy compared with the original torch code.

**Comment by zhiqiangwan** [comment]
Thank you for your reply. Did you find out what lead to the lower accuracy in the previous version?

**Comment by chingyaoc** [comment]
Some steps of image normalization went wrong, everything is good after correcting it!
